mesonbuild/compilers/c.py

```diff
@@ -0,0 +1,996 @@
+            if mode == 'link':
+                # Add link flags needed to find dependencies
+                args += d.get_link_args()
+        # Select a CRT if needed since we're linking
+        if mode == 'link':
+            args += self.get_linker_debug_crt_args()
+        # Read c_args/cpp_args/etc from the cross-info file (if needed)
+        args += self.get_cross_extra_flags(env, link=(mode == 'link'))
+        if not self.is_cross:
+            if mode == 'preprocess':
+                # Add CPPFLAGS from the env.
+                args += env.coredata.external_preprocess_args[self.language]
+            elif mode == 'compile':
+                # Add CFLAGS/CXXFLAGS/OBJCFLAGS/OBJCXXFLAGS from the env
+                args += env.coredata.external_args[self.language]
+            elif mode == 'link':
+                # Add LDFLAGS from the env
+                args += env.coredata.external_link_args[self.language]
+        args += self.get_compiler_check_args()
+        # extra_args must override all other arguments, so we add them last
+        args += extra_args
+        return args
+
+    def compiles(self, code, env, extra_args=None, dependencies=None, mode='compile'):
+        args = self._get_compiler_check_args(env, extra_args, dependencies, mode)
+        # We only want to compile; not link
+        with self.compile(code, args.to_native(), mode) as p:
+            return p.returncode == 0
+
+    def _links_wrapper(self, code, env, extra_args, dependencies):
+        "Shares common code between self.links and self.run"
+        args = self._get_compiler_check_args(env, extra_args, dependencies, mode='link')
+        return self.compile(code, args.to_native())
+
+    def links(self, code, env, extra_args=None, dependencies=None):
+        with self._links_wrapper(code, env, extra_args, dependencies) as p:
+            return p.returncode == 0
+
+    def run(self, code, env, extra_args=None, dependencies=None):
+        if self.is_cross and self.exe_wrapper is None:
+            raise CrossNoRunException('Can not run test applications in this cross environment.')
+        with self._links_wrapper(code, env, extra_args, dependencies) as p:
+            if p.returncode != 0:
+                mlog.debug('Could not compile test file %s: %d\n' % (
+                    p.input_name,
+                    p.returncode))
+                return RunResult(False)
+            if self.is_cross:
+                cmdlist = self.exe_wrapper + [p.output_name]
+            else:
+                cmdlist = p.output_name
+            try:
+                pe, so, se = Popen_safe(cmdlist)
+            except Exception as e:
+                mlog.debug('Could not run: %s (error: %s)\n' % (cmdlist, e))
+                return RunResult(False)
+
+        mlog.debug('Program stdout:\n')
+        mlog.debug(so)
+        mlog.debug('Program stderr:\n')
+        mlog.debug(se)
+        return RunResult(True, pe.returncode, so, se)
```

Comment:
__[flake8]__
* [F405] 'RunResult' may be undefined, or defined from star imports: .compilers

<!-- Comment by [SideCI](https://sideci.com) -->